### DISCUSSION TITLE
feat: add a mode in which it will populate the indexed images' EXIF data with tags (being implemented as a separate tool, `rtag`)

### DISCUSSION THREAD

**Discussion by yurijmikhalevich** [yurijmikhalevich/rclip#254] [Ideas]
### UPD: this feature is being worked on and will be released as a separate tool, [rtag](https://github.com/yurijmikhalevich/rtag). You can follow the development by starring and watching the [rtag repository](https://github.com/yurijmikhalevich/rtag). This Issue will be closed once rtag v1 is released.

Tags should be taken from a fixed, pre-defined set. The ImageNet competition tag list is a good place to start, wikientities is another one.

The v1 of the feature should look like this from the user perspective:
```
$ rclip tag
This operation will overwrite the tags stored in your images' EXIF metadata. This will change the image files and will make you loose any existing keywords stored on your images.
Proceed? y/n y
/photos [#####]
```

For v1, it should write the tags from the list of the ImageNet-1k tags, based on a certain threshold (should be determined empirically).

Interested user: https://www.youtube.com/watch?v=MsTgYdOpgcQ&lc=Ugzl8qVAiVsek_2TwoZ4AaABAg

**Comment by yburkouski** [comment]
If it will add (or override?) data to "Tags'' then it will be fine! I don't know, maybe someone is using that field for photo “catalogization”, so adding to the user custom tags could also be an option (but in that case we should not duplicate data during second run, could be complicated depending on implementation). And I think it's better to have a parameter for the number of tags to put, 10 by default.
![image](https://github.com/yurijmikhalevich/rclip/assets/16322952/f18a0bef-cc85-443e-be73-81c7cfa05d80)

**Comment by yurijmikhalevich** [comment]
Hi @yburkouski! Thank you for your interest and for sharing your thoughts. These are all valid points that are being considered. For the most part, you have literally described the design I have in mind :)

> And I think it's better to have a parameter for the number of tags to put, 10 by default.

Having a confidence threshold instead of a hard number of tags will work better because it will allow getting in all high-confidence labels and ignoring everything else, even if there are only 3 high-confidence labels.

**Comment by stubkan** [comment]
Why not sidecar text files?  Sidecars are normal for many image viewers / managers and dont touch the images themselves.  Wouldn't be hard for python to append tags to a text file that has the same name as the images.  You could also write the confidence levels of the tags into these files.

Then if one was inclined, there are then, many tools that work with sidecars to add or modify image metadata.

**Comment by yurijmikhalevich** [comment]
@stubkan, that's an interesting suggestion. So, you are saying that you'd rather not let the tool modify your images. Right? I'll give it a closer look.

**Comment by yurijmikhalevich** [comment]
@stubkan, the XMP sidecar support will be implemented as a part of [rtag](https://github.com/yurijmikhalevich/rtag/issues/1)

**Comment by rebelmachina** [comment]
> Why not sidecar text files? Sidecars are normal for many image viewers / managers and dont touch the images themselves. Wouldn't be hard for python to append tags to a text file that has the same name as the images. You could also write the confidence levels of the tags into these files.
> 
> Then if one was inclined, there are then, many tools that work with sidecars to add or modify image metadata.

What's a sidecar? just curious never heard of that. So you want to save a metadata file and not modify the image ?

**Comment by yburkouski** [comment]
> What's a sidecar? just curious never heard of that. So you want to save a metadata file and not modify the image ?

@rebelzion need assistance to open google.com? ;) "sidecar text file"
